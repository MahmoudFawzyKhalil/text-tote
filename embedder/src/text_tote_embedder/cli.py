from .configuration import appconfig

def app():
    print(appconfig)
