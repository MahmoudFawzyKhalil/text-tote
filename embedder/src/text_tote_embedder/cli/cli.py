def app():
    print("Cli App")
