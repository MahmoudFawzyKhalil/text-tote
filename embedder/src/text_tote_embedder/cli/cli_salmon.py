from pathlib import Path
from typing import List

import typer

from salmon_search.schemas import Resource, ChunkRecord

app = typer.Typer(rich_markup_mode="rich")


@app.command()
def init():
    """
    Initialize the sqlite database. Must be called before using any other command.
    """
    from salmon_search import db
    import click

    try:
        db.create_db()
    except db.DbAlreadyExistsException:
        raise click.ClickException("Database already exists.")


@app.command()
def index(url: str = typer.Option(None, help="URL of the resource to index."),
          file: Path = typer.Option(None, help="Path to a file with URLs of the resources to index."),
          playlist: str = typer.Option(None, help="YouTube playlist ID to index."),
          apikey: str = typer.Option(None, help="YouTube API key. Must specify if --playlist is used."),
          maxvideos: int = typer.Option(50, help="Maximum number of playlist videos to index")):
    """
    Index a resource e.g. an article, or a YouTube video.

    salmon index --url https://en.wikipedia.org/wiki/Salmon
    salmon index --file urls.txt
    """
    import click

    from salmon_search import db
    db.update_vss_index()

    if url is not None:
        resources = [index_url(url)]
    elif file is not None:
        resources = index_file(file)
    elif playlist is not None:
        import rich
        resources = index_playlist(playlist, apikey, maxvideos)

        if apikey is None:
            raise click.ClickException("--apikey must be specified if --playlist is used.")
    else:
        raise click.ClickException("Either --url or --file must be specified.")
    db.update_vss_index()
    print_resource_table(resources)


def index_playlist(playlist_id: str, apikey: str, maxvideos: int) -> list[Resource]:
    from salmon_search import resources
    from salmon_search import db

    json_dict = get_playlist_items(playlist_id, apikey)

    videos = []
    extract_videos(json_dict, videos)

    next_page_token = json_dict["nextPageToken"]
    while next_page_token and len(videos) < maxvideos:
        json_dict = get_playlist_items(playlist_id, apikey, next_page_token)
        extract_videos(json_dict, videos)
        next_page_token = json_dict.get("nextPageToken")

    created_resources = []
    for video in videos:
        url = f"https://youtu.be/{video['videoId']}"
        valid = validate_url(url)
        if not valid:
            continue
        resource: Resource = resources.create_youtube_video_resource(
            url,
            video["title"],
            video["description"])
        created_resources.append(resource)
        db.save_resource(resource)

    return created_resources


def get_playlist_items(playlist_id: str, apikey: str, next_page_token: str = None) -> dict:
    import requests
    url = "https://content-youtube.googleapis.com/youtube/v3/playlistItems"
    params = {
        "maxResults": 50,
        "part": "id,snippet",
        "playlistId": playlist_id,
        "key": apikey
    }
    if next_page_token:
        params["pageToken"] = next_page_token

    res = requests.get(url, params=params)
    return res.json()


def extract_videos(json_dict, videos):
    for item in json_dict["items"]:
        video = {
            "videoId": item["snippet"]["resourceId"]["videoId"],
            "title": item["snippet"]["title"],
            "description": item["snippet"]["description"]
        }
        videos.append(video)


def index_file(file: Path) -> list[Resource]:
    resources: list[Resource] = []
    with open(file, 'r') as f:
        for line in f:
            line = line.strip()
            if line:
                resources.append(index_url(line))
    return resources


def index_url(url: str) -> Resource | None:
    from salmon_search import db
    from salmon_search import resources
    import rich

    valid = validate_url(url)
    if not valid:
        return None
    try:
        resource: Resource = resources.create_resource(url)
        db.save_resource(resource)
    except Exception as e:
        rich.print(f"[red]Error saving resource {url} to database. Skipping. Error:{str(e)}[/red]")
        return None
    return resource


def print_resource_table(resources: list[Resource]):
    import rich.table as table
    import rich

    console = rich.console.Console()

    tbl: table.Table = table.Table("id",
                                   table.Column("url"),
                                   "title",
                                   "#chunks",
                                   title="Resources")
    for resource in [r for r in resources if r is not None]:
        tbl.add_row(str(resource.id), resource.url, resource.title, str(len(resource.chunks)))

    console.print(tbl)


def print_matches_table(matches: list[ChunkRecord]):
    import rich.table as table
    import rich

    console = rich.console.Console()

    tbl: table.Table = table.Table("#",
                                   "dist",
                                   "rid",
                                   table.Column("resource_title", max_width=40),
                                   "resource_url",
                                   # "most_similar_chunk",
                                   "cid",
                                   title="Matches")
    for i, chunk in enumerate(matches):
        tbl.add_row(
            str(i),
            f"{chunk.distance:.2f}",
            str(chunk.resource_id),
            chunk.resource_title,
            chunk.resource_url,
            # chunk.chunk,
            str(chunk.chunk_id)
        )

    console.print(tbl)


@app.command()
def search(query: str = typer.Argument(..., help="Search query."),
           n: int = typer.Option(100, help="Number of text chunks to compare the query to."),
           o: str = typer.Option("table", help="Output format. One of: table, json.")):
    """
    Semantically search resources.

    salmon search "Where do Salmons live?"
    """
    import rich
    from salmon_search import db
    from salmon_search import embeddings

    query_embedding = embeddings.encode(query)
    results: list[ChunkRecord] = db.get_most_similar_articles_based_on_n_chunks(n, query_embedding)
    if o == "table":
        print_matches_table(results)
    elif o == "json":
        rich.print_json(data=results)


@app.command()
def delete(resource_ids: List[int] = typer.Argument(..., help="resource_id for resource to delete and un-index.")):
    """
    Delete a resource from the index.

    salmon delete 1
    """
    from salmon_search import db

    deleted_resources = []
    for rid in resource_ids:
        resource = db.delete_resource(rid)
        deleted_resources.append(resource)
    print_resource_table(deleted_resources)


@app.command()
def get(rid: int = typer.Option(None, help="Resource ID of resource to get."),
        cid: int = typer.Option(None, help="Chunk ID of chunk to get")):
    """
    Get a resource or chunk.

    salmon get --rid 1
    """
    import click
    import rich

    from salmon_search import db

    if rid is not None:
        resource = db.get_resource(rid)
        print_resource_table([resource])
    elif cid is not None:
        chunk = db.get_chunk(cid)
        rich.print(chunk)
    else:
        raise click.ClickException("Either --rid or --cid must be specified.")


def validate_url(url) -> bool:
    import validators
    from salmon_search import db
    import rich

    if not validators.url(url):
        rich.print(f"[red]Invalid URL: {url}, skipping.[/red]")
        return False
    if db.resource_exists_by_url(url):
        # rich.print(f"[green]Resource already exists: {url}, skipping.[/green]")
        return False
    return True


def main_cli():
    app()
