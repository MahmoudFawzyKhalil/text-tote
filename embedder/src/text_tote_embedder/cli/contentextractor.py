"""
Extracts the main content of a URL (article or YouTube video) as stripped-down HTML and plain text.
"""

import requests
from readability import Document
from bs4 import BeautifulSoup

from dataclasses import dataclass

@dataclass
class Content:
    title: str
    summary: str
    clean_text: str
    html: str
    url: str
    source: str
    source_url: str
    source_name: str
    source_logo: str
    source_favicon: str
    source_description: str
    source_language: str
    source_country: str
    source_category: str
    source_subcategory: str
    source_published_date: str
    source_last_updated_date: str
    source_author: str
    source_author_url: str
    source_author_image: str
    source_author_bio: str
    source_author_location: str
    source_author_website: str
    source_author_social_links: str
    source_author_social_media: str
    source_author_social_media_platforms: str
    source_author_social_media_accounts: str
    source_author_social_media_accounts_urls: str
    source_author_social_media_accounts_platforms: str
#
def extract_content(url) -> Content:
#     """
#     Extracts the main content of a URL (article or YouTube video) as stripped-down HTML and plain text.
#
#     Args:
#         url (str): The URL to extract content from.
#
#     Returns:
#         tuple: A tuple containing the title and clean text summary of the URL content.
#     """

    url = 'https://www.kdnuggets.com/sql-simplified-crafting-modular-and-understandable-queries-with-ctes'

    # Fetch the HTML content of the URL
    response = requests.get(url)
    html = response.content

    # Parse the HTML using readability
    doc = Document(html)

    # Extract the main content
    title = doc.title()
    summary = doc.summary(html_partial=True)

    print()
    print("Title:", title)

    # Convert HTML summary to plain text

    soup = BeautifulSoup(summary, 'html.parser')
    clean_text = soup.get_text(separator='\n')
    print("Clean Text:", clean_text)
