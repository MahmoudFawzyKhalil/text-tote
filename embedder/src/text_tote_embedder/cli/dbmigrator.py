import os
import re
import sqlite3
from os import path


def migrate_db(db_connection: sqlite3.Connection):
    cursor = db_connection.cursor()

    try:
        script_directory = path.dirname(path.abspath(__file__))
        migrations_directory = path.join(script_directory, 'migrations')

        sql_scripts = sorted(os.listdir(migrations_directory))

        create_schema_version_table_if_not_exist(cursor)

        # Fetch the current schema version
        current_version = get_current_schema_version(cursor)
        print(f"Current Schema Version: {current_version}")

        for sql_script in sql_scripts:
            if sql_script.endswith('.sql'):
                file_path = path.join(migrations_directory, sql_script)

                version = extract_version(sql_script)
                if version > current_version:
                    execute_sql_file(cursor, file_path)
                    update_schema_version(cursor, version)
                    print(f"Schema version updated to: {version}")

        db_connection.commit()
        print("Database migrated successfully.")

    except Exception as e:
        print(f"Error executing SQL scripts: {e}")

    finally:
        db_connection.close()


def create_schema_version_table_if_not_exist(cursor):
    cursor.execute("CREATE TABLE IF NOT EXISTS schema_version (version integer PRIMARY KEY)")
    cursor.execute("SELECT * FROM schema_version")
    single_row_already_exists = cursor.fetchone() is not None
    if single_row_already_exists:
        return
    else:
        cursor.execute("INSERT INTO schema_version (version) VALUES (0)")


def extract_version(file_name: str) -> int:
    match = re.match(r'V(\d+)', file_name)
    if match:
        return int(match.group(1))
    else:
        raise Exception(f"Invalid file name: {file_name}")


def execute_sql_file(cursor, file_path: str):
    with open(file_path, 'r') as sql_file:
        print(f"Executing SQL script: {file_path}")
        sql_script = sql_file.read()
        cursor.executescript(sql_script)


def get_current_schema_version(cursor):
    # Fetch the current schema version from the database or return a default value if the table is not present
    cursor.execute("SELECT version FROM schema_version")
    result = cursor.fetchone()
    return result[0] if result else 0


def update_schema_version(cursor, version):
    # Update the schema_version table with the new version
    cursor.execute("UPDATE schema_version SET version = ?", (version,))
