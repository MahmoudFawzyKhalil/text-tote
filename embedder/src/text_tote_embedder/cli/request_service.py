import logging

import requests

HEADERS = {
    "User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.0.0 Safari/537.36"
}

TIMEOUT = (10, 10)


def get(url: str) -> bytes:
    # TODO check out more webscraping tips to avoid detection
    response = requests.get(url=url, headers=HEADERS, timeout=TIMEOUT, allow_redirects=True)
    logging.info(f"GET url={url}, response_code={response.status_code}")

    if response.status_code != 200:
        raise Exception(f"GET request failed. url={url}, response_code={response.status_code}")
    return response.content
