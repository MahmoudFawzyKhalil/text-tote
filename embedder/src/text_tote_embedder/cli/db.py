import os
import sqlite3

import sqlite_vss

from text_tote_embedder.cli import dbmigrator
from text_tote_embedder.configuration import config


def create_db() -> None:
    """
    Create the database if it doesn't exist.
    """
    if os.path.exists(config.db_path):
        return

    uri = f"file:{config.db_path}?mode=rwc"
    conn = sqlite3.connect(uri, uri=True)
    _load_extensions(conn)


def migrate_db() -> None:
    """
    Migrate the database objects to the latest version. This will create the database if it doesn't exist.
    """
    create_db()
    db_connection = _create_connection()
    dbmigrator.migrate_db(db_connection)


def _create_connection() -> sqlite3.Connection:
    uri = f"file:{config.db_path}?mode=rw"
    db_connection = sqlite3.connect(uri, uri=True)
    _load_extensions(db_connection)
    return db_connection


def _load_extensions(db_connection: sqlite3.Connection) -> None:
    db_connection.enable_load_extension(True)
    sqlite_vss.load(db_connection)
