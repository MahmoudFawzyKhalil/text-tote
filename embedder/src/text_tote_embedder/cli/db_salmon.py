import os
import sqlite3

import numpy as np
import sqlite_vss

from text_tote_embedder import embeddings
from .schemas import ChunkRecord, Resource, chunk_record_factory

SALMON_DIR = os.getcwd()
DB_PATH = os.path.join(SALMON_DIR, 'salmon.db')
# create ~/salmon directory if it doesn't exist
os.makedirs(SALMON_DIR, exist_ok=True)


def add(a, b):
    return a + b


def create_connection(create: bool = False):
    mode = f"rw{create and 'c' or ''}"
    uri = f"file:{DB_PATH}?mode={mode}"
    conn = sqlite3.connect(uri, uri=True)
    conn.enable_load_extension(True)
    sqlite_vss.load(conn)
    return conn


class DbAlreadyExistsException(Exception):
    pass


def create_db():
    if os.path.exists(DB_PATH):
        raise DbAlreadyExistsException('Database already exists.')

    conn = create_connection(True)
    conn.execute('''
        CREATE TABLE if NOT EXISTS resources (
            id INTEGER PRIMARY KEY,
            url TEXT,
            title TEXT
        );
    ''')

    conn.execute('''
        CREATE TABLE if NOT EXISTS chunks (
            id INTEGER PRIMARY KEY,
            chunk TEXT,
            embedding BLOB,
            resource_id INTEGER,
            FOREIGN KEY (resource_id) REFERENCES resources (id)
        );
    ''')

    conn.execute(f'''
        CREATE VIRTUAL TABLE IF NOT EXISTS vss_chunks USING vss0(
        chunk_embedding({embeddings.VECTOR_SIZE});
    );
    ''')

    # Create empty chunk embedding to avoid issue with creating empty vss_table
    zero_array = np.zeros(embeddings.VECTOR_SIZE)
    conn.execute('''
        INSERT INTO vss_chunks (chunk_embedding)
            VALUES (?)
            ''', [zero_array])

    conn.commit()
    conn.close()


def resource_exists_by_url(url: str):
    conn = create_connection()
    cursor = conn.cursor()
    cursor.execute('''
        SELECT id FROM resources WHERE url = ?
    ''', [url])
    result = cursor.fetchone()
    cursor.close()
    conn.close()
    return result is not None


def save_resource(resource: Resource):
    conn = create_connection()
    cursor = conn.cursor()

    cursor.execute('''
        INSERT INTO resources (url, title)
        VALUES (?, ?)
    ''', (resource.url, resource.title))
    resource_id = cursor.lastrowid
    for chunk, embedding in zip(resource.chunks, resource.embeddings):
        cursor.execute('''
            INSERT INTO chunks (chunk, embedding, resource_id)
            VALUES (?, ?, ?)
        ''', [chunk, embedding, resource_id])

    resource.id = resource_id
    cursor.close()
    conn.commit()
    conn.close()


def update_vss_index():
    conn = create_connection()
    conn.execute('''
            INSERT INTO vss_chunks (rowid, chunk_embedding)
                SELECT c.rowid, c.embedding
                FROM chunks c
            WHERE c.rowid > COALESCE((SELECT MAX(v.rowid) FROM vss_chunks v), 0)
        ''')

    conn.commit()
    conn.close()


def get_most_similar_articles_based_on_n_chunks(n: int, query_embedding: np.ndarray) -> list[ChunkRecord]:
    conn = create_connection()
    cursor = create_cursor_with_row_factory(conn)
    cursor.execute('''
    SELECT MIN(distance) dist, c.rowid, c.chunk, r.rowid, r.title, r.url
    FROM vss_chunks v
    JOIN chunks c ON c.rowid = v.rowid
    JOIN resources r ON r.id = c.resource_id
    WHERE vss_search(
      v.chunk_embedding,
      vss_search_params(
        ?,
        ?
      )
    )
    GROUP BY r.rowid
    ORDER BY dist
    ''', [query_embedding, n])
    results: list[ChunkRecord] = cursor.fetchall()
    cursor.close()
    conn.close()
    return results


def get_top_n_chunks(n: int, query_embedding: np.ndarray) -> list[ChunkRecord]:
    conn = create_connection()
    cursor = create_cursor_with_row_factory(conn)
    cursor.execute('''
    SELECT distance dist, c.rowid, c.chunk, r.rowid, r.title, r.url
    FROM vss_chunks v
    JOIN chunks c ON c.rowid = v.rowid 
    JOIN resources r ON r.id = c.resource_id
    WHERE vss_search(
      v.chunk_embedding,
      vss_search_params(
        ?1,
        ?2
      )
    )
    ORDER BY dist
    ''', [query_embedding, n])
    results: list[ChunkRecord] = cursor.fetchall()
    cursor.close()
    conn.close()
    return results


def create_cursor_with_row_factory(conn):
    cursor = conn.cursor()
    cursor.row_factory = chunk_record_factory
    return cursor


def get_resource(resource_id: int) -> Resource:
    conn = create_connection()
    cursor = conn.cursor()

    def row_factory(cursor, row):
        return Resource.from_args(*row)

    cursor.row_factory = row_factory

    cursor.execute('''
    SELECT rowid, url, title
    FROM resources
    WHERE rowid = ?
    ''', [resource_id])
    return cursor.fetchone()


def delete_resource(resource_id: int):
    conn = create_connection()
    cursor = conn.cursor()
    resource = get_resource(resource_id)
    cursor.execute(
        '''
        DELETE FROM vss_chunks
        WHERE rowid IN (
            SELECT rowid
            FROM chunks
            WHERE resource_id = ?
        )
        ''',
        [resource_id]
    )
    cursor.execute(
        '''
        DELETE FROM chunks
        WHERE resource_id = ?
        ''',
        [resource_id]
    )
    cursor.execute(
        '''
        DELETE FROM resources
        WHERE rowid = ?
        ''', [resource_id]
    ).fetchone()
    cursor.close()
    conn.commit()
    conn.close()
    return resource


def get_chunk(chunk_id: int) -> str:
    conn = create_connection()
    cursor = conn.cursor()
    chunk = cursor.execute(
        '''
        SELECT chunk
        FROM chunks
        WHERE rowid = ?
        ''',
        [chunk_id]
    ).fetchone()
    if chunk:
        return chunk[0]
    else:
        return f'No chunk with ID {chunk_id}'
