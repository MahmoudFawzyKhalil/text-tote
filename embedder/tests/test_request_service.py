import approvaltests

from text_tote_embedder.cli import request_service


def test_retrieves_content():
    # Arrange
    url = 'https://en.wikipedia.org/wiki/Java_(programming_language)'

    # Act
    content: bytes = request_service.get(url)

    # Assert
    approvaltests.verify(content)


def test_fails_when_status_code_is_not_200():
    # Arrange
    url = 'https://en.wikipedia.org/wiki/Non-existent-page'

    # Act
    try:
        request_service.get(url)
    except Exception as e:
        # Assert
        approvaltests.verify(e)
