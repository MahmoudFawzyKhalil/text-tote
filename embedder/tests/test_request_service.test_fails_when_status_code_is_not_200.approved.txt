GET request failed. url=https://en.wikipedia.org/wiki/Non-existent-page, response_code=404
